### DISCUSSION TITLE
Module for Custom Priors and Distributions

### DISCUSSION THREAD

**Discussion by steven-struglia** [google/lightweight_mmm#322] [Ideas]
Hello everyone,

I've been working on setting up custom priors for the lightweight_mmm model, specifically for the hill-adstock model. After some effort, I have finally figured out a way to re-assign the distributions and parameters effectively. I'm planning to create a Pull Request (PR) to contribute this to the open-source effort, but I wanted to share the implementation here for immediate use for those looking for a solution.

Custom Priors Implementation
Here is how you can define custom priors using different distributions:

```
custom_priors = {
    "lag_weight": dist.Uniform(low=0.0, high=1.0),
    "half_max_effective_concentration": dist.Uniform(low=0.0, high=10.0),
    "slope": dist.Gamma(concentration=1.0, rate=1.0),
    "intercept": dist.Normal(loc=0.0, scale=2.0),
    "coef_trend": dist.Normal(loc=0.0, scale=1.0),
    "expo_trend": dist.Uniform(low=0.5, high=1.5),
    "sigma": dist.Gamma(concentration=1.0, rate=1.0),
    "gamma_seasonality": dist.Normal(loc=0.0, scale=1.0),
    "weekday": dist.Normal(loc=0.0, scale=0.5),
    "coef_extra_features": dist.Normal(loc=0.0, scale=1.0)
}

class CustomLightweightMMM(lightweight_mmm.LightweightMMM):
    def __init__(self, model_name="hill_adstock", custom_priors=None):
        super().__init__(model_name)
        self.custom_priors = custom_priors or {}

    def fit(self, media, media_prior, target, extra_features=None, **kwargs):
        default_priors = dict(_get_default_priors())
        transform_default_priors = dict(_get_transform_default_priors()[self.model_name])

        # Update the priors with custom values if provided
        for prior_name, custom_prior in self.custom_priors.items():
            if prior_name in default_priors:
                default_priors[prior_name] = custom_prior
            if prior_name in transform_default_priors:
                transform_default_priors[prior_name] = custom_prior

        # Remove geo-specific priors
        geo_specific_priors = {"coef_seasonality"}
        for prior in geo_specific_priors:
            if prior in default_priors:
                del default_priors[prior]
            if prior in transform_default_priors:
                del transform_default_priors[prior]

        self.custom_priors = immutabledict.immutabledict({**default_priors, **transform_default_priors})
        super().fit(media, media_prior, target, extra_features, custom_priors=self.custom_priors, **kwargs)

custom_mmm = CustomLightweightMMM(model_name="hill_adstock", custom_priors=custom_priors)
```

From here, you can run the fit() method on custom_mmm as you would with the normal model, and it will use the custom priors defined in the dictionary above. Please let me know if there's interest in having this as part of the package itself for people to use consistently. I'll be happy to proceed with creating a PR for this contribution.

**Comment by anjaliagarwal88** [comment]
Hi Steven,
Same as you, I've also been working on setting up custom priors for the lightweight_mmm model, specifically for the hill-adstock model. 
In my case, I know the prior contribution percentages of the 5 channels. I know that their sum should be let's say 10. would you by any chance, know how to set custom priors in this case? I tried quiet a few things like tweaking intercept values with different half-normal and normal with different loc and scale values individually, but even after using different combinations , I am still unable to get the desired contribution percentages.
Below is the code that I have been trying:


loc_range = range(250, 1000, 200)  
scale_range = range(100, 20, 3)  

# Loop through each combination of loc and scale
for loc in loc_range:
    for scale in scale_range:
        #intercept_prior = dist.Normal(loc=loc, scale=scale)
        intercept_prior = dist.HalfNormal(scale=scale)

        # You can print or store the intercept_prior as needed
        #print(f"Intercept Prior with loc={loc} and scale={scale}: {intercept_prior}")
        print(f"Intercept Prior with scale={scale}: {intercept_prior}")
        
        mmm.fit(
            media=media_data_train,
            media_names=media_names,
            media_prior=costs,
            target=target_train,
            extra_features=extra_features_train,
            number_warmup=number_warmup,
            number_samples=number_samples,
            seasonality_frequency=365,
            weekday_seasonality=True,
            seed=SEED,
            custom_priors={
                "intercept": intercept_prior,
                "weekday": weekday_prior
                            }
            )

Please suggest if you have any insights regarding this.
